Integration Test Framework

justenwalker opened the pull request:
For #67 

Still a work in progress, but it's finally working for me in TravisCI.  

- There is a [README.md](https://github.com/justenwalker/containerbuddy/blob/integration_tests/integration_tests/README.md) in the integration_tests folder which describes how it operates and how to make tests.
- I've made an example test for SIGTERM. It's small golang application called [test_probe](https://github.com/justenwalker/containerbuddy/tree/integration_tests/integration_tests/fixtures/test_probe). It can be extended to support other tests easily - reducing the amount of bash wizardry needed to write tests.
- Added a test for #74 (but forced it to pass)

tgross commented on integration_tests/README.md:
```diff
@@ -0,0 +1,56 @@
+
+Tests Directory: `integration_tests/tests`
+
+- Folders in the tests directory will be run as tests
```
I think it's clear to me and you but we should include a note here that these tests are run in the context of the testing host (i.e. TravisCI or your local environment), not inside a container.

tgross replied:
This also implies (as the TravisCI config file says) that we need golang on the test-running host. We should make sure to document those assumptions as well. (Or do we want to run the tests themselves in a container?)

justenwalker replied:
Actually, we don't technically need Golang on the host environment.  the compose script can bring up a golang container to do testing instead of doing it locally.

Really I added the go language to `.travis.yml` because,  by default, TravisCI uses RVM and Ruby.  It just makes it clear which language we are targeting - has no actual effect the way we are using it.

tgross replied:
Got it, understood.

tgross commented on test.sh:
```diff
@@ -0,0 +1,160 @@
+    done
+}
+
+case ${1:-"test"} in
```
Maybe we should have the option to run single tests here? I could imagine a case where if a developer is trying to fix a broken test that having to wait for the whole rig to build would be painful. That does mean that:

> *Note*: Since fixtures are created in alpha order, they can have FROM directives for previously created images

probably wouldn't work. But maybe that's a good thing?

justenwalker replied:
Easy enough.  Fixtures are created before tests.  I can make it run a specific test, but still build all fixtures.

justenwalker commented:
- Added option to run an individual integration test
- Made `CONTAINERBUDDY_BIN` available to docker-compose
- Altered `test_sighup_deadlock` to increase the chance of finding the issue
- Updated `README` to expand on how to run tests

tgross commented:
This looks great, nice work, @justenwalker
